passwordstore: Support subkey creation and update

mluzarreta opened the pull request:
##### SUMMARY

passwordstore lookup - Add the ability to create and update subkeys in password files.

##### ISSUE TYPE

- Feature Pull Request

##### COMPONENT NAME

passwordstore lookup

##### ADDITIONAL INFORMATION

Here some command lines to  illustrate the feature:

```
# Create a new 'foo' subkey, in an new password store file (in this case, main password is empty)
❯ ansible -m debug -i localhost, -a "msg={{ lookup('community.general.passwordstore', 'example/test', create=true, subkey='foo') }}" localhost
localhost | SUCCESS => {
    "msg": "u6V31gj3d0i3lLbO"
}
# Replace the 'foo' subkey and keep a backup trace in the password file
❯ ansible -m debug -i localhost, -a "msg={{ lookup('community.general.passwordstore', 'example/test', create=true, subkey='foo', overwrite=true, backup=true) }}
" localhost
localhost | SUCCESS => {
    "msg": "8jRMhmn9bE-Ra:Yh"
}
# Replace the main password
❯ ansible -m debug -i localhost, -a "msg={{ lookup('community.general.passwordstore', 'example/test', create=true, overwrite=true) }}" localhost
localhost | SUCCESS => {
    "msg": "Mjol:uhPxIiGcchM"
}
# Display the content of the password file
❯ pass example/test 
Mjol:uhPxIiGcchM

foo: 8jRMhmn9bE-Ra:Yh
lookup_pass: First generated by ansible on 01/10/2024 11:37:25
lookup_pass: old subkey password foo was u6V31gj3d0i3lLbO (Updated on 01/10/2024 11:38:03)

```

felixfontein commented on changelogs/fragments/8952-password-store-lookup-create-subkey-support.yml:
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+    - passwordstore lookup - add subkey creation/update support
```
```suggestion
    - passwordstore lookup plugin - add subkey creation/update support (https://github.com/ansible-collections/community.general/pull/8952).
```

felixfontein reviewed (commented):
Thanks for your contribution! A few first comments are below:

The option actually isn't added to `DOCUMENTATION` and `self.paramvals` (line 554 and following ones), so you cannot actually use it.

russoz commented on plugins/lookup/passwordstore.py:
```diff
@@ -525,11 +577,17 @@ def run(self, terms, variables, **kwargs):
-                        with self.opt_lock('write'):
+                    if self.paramvals['overwrite']:
+                        if self.paramvals['subkey'] == 'password':
+                          with self.opt_lock('write'):
                             result.append(self.update_password())
+                        elif not self.passdict.get(self.paramvals['subkey']):  # password exists but not the subkey
```
Indentation is off.
```suggestion
                            with self.opt_lock('write'):
                                result.append(self.update_password())
                        elif not self.passdict.get(self.paramvals['subkey']):  # password exists but not the subkey
```

russoz commented on plugins/lookup/passwordstore.py:
```diff
@@ -525,11 +577,17 @@ def run(self, terms, variables, **kwargs):
+                            result.append(self.get_passresult())
                     else:
-                        result.append(self.get_passresult())
+                            result.append(self.get_passresult())
```
The indentation appears to be off here as well, please review.

mluzarreta commented:
> Thanks for your contribution! A few first comments are below:
> 
> The option actually isn't added to `DOCUMENTATION` and `self.paramvals` (line 554 and following ones), so you cannot actually use it.

@felixfontein I added some details in the documentation but this PR does not add any parameter, it just allows the creation/update of a subkey using the existing parameters (subkey, create, overwrite, backup).

@felixfontein @russoz I rewrote the commit with your suggestions. Thank you for the review :smile:

felixfontein commented:
> The option actually isn't added to `DOCUMENTATION` and `self.paramvals` (line 554 and following ones), so you cannot actually use it.

It seems I totally overlooked that the option already exists. Sorry for that :)

mluzarreta commented:
FYI I've just corrected a regression:

```
-                    elif not self.passdict.get(self.paramvals['subkey']):  # password exists but not the subkey
+                    elif self.paramvals["subkey"] != "password" and not self.passdict.get(self.paramvals['subkey']):  # password exists but not the subkey
```

russoz reviewed (approved):
LGTM

felixfontein commented:
If nobody objects, I'll merge this in a week.

felixfontein commented:
@mluzarreta thanks for your contribution!
@russoz thanks for reviewing!
